spring-aop-2/src/test/java/com/baeldung/exceptionhandling/ExceptionHandlingAspectTest.java

```diff
@@ -0,0 +1,61 @@
+package com.baeldung.exceptionhandling;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.List;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
+
+@SpringBootTest
+public class ExceptionHandlingAspectTest {
+
+    @Autowired
+    private PaymentService paymentService;
+
+    private ListAppender<ILoggingEvent> listAppender;
+
+    @BeforeEach
+    public void setup() {
+        Logger logger = (Logger) LoggerFactory.getLogger(ExceptionHandlingAspect.class);
+
+        listAppender = new ListAppender<>();
+        listAppender.start();
+
+        logger.addAppender(listAppender);
+    }
+
+    @Test
+    public void givenInvalidPaymentDetails_whenValidationFails_thenAspectLogsException() {
+        // When
+        assertThatThrownBy(() -> paymentService.validatePaymentDetails("")).isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Payment details cannot be null or empty.");
+
+        // Then
+        List<ILoggingEvent> logsList = listAppender.list;
+        assertThat(logsList).isNotEmpty();
+        assertThat(logsList).extracting(Object::toString)
```

Comment:
We are still not showing that the Exception is logged:
```
logger.error("Exception occurred: {}", ex.getMessage(), ex);
```

The last parameter `ex` is untested at the moment

Reply:
The last parameter "ex" is not actually needed. I have removed it in the latest commit. 

`assertThat(logsList).extracting(Object::toString)
            .containsExactly("[ERROR] Exception occurred: Payment details cannot be null or empty.");`

This assertion captures the logs and checks whether the exception that actually occurred matches the expected log.

Reply:
Sorry but I have to disagree here. It's essential when logging exceptions to preserve the stack trace - can we find a way to test this please?

Reply:
I agree that preserving the stacktrace is important. But I can't think of a way to create an assertion for that. One way can be to actually assert each line of the stacktrace printed but that will be too clumsy. 

On a side note, isn't it implicit that if the code has reached the error logger and thrown an error message that we have already asserted, then the stack trace is already tested? Let me know if you think otherwise and has a solution to test.

Reply:
Ok I'll show you how to do this later when I get some time

Reply:
And no it's not implicit. The test would still be green if I replaced the ex parameter with a dummy new Exception()

Reply:
Yup! Got it.

Reply:
You can do something like:
```
        assertThat(logsList)
            .extracting(event -> event.getThrowableProxy().getMessage())
            .containsExactly("Payment processing failed due to invalid details.");
```

or alternatively, wire a mock/spy Logger into the ExceptionHandlingAspect and use Mockito to verify